eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -1247,6 +1301,168 @@ submitFailingTxM tx mkExpectedFailures = do
+  ImpTestM era (Block BHeaderView era)
+withTxsInBlock act = do
+  txs <- simulateThenRestore $ impTxsFrom act
+  submitBlock txs
+
+-- | Gather all the txs submitted by @act@ and resubmit them as a block
+-- that's expected to fail with the given predicate failures.
+withTxsInFailingBlock ::
+  ( HasCallStack
+  , ShelleyEraImp era
+  ) =>
+  ImpTestM era () ->
+  NonEmpty (PredicateFailure (EraRule "BBODY" era)) ->
+  ImpTestM era ()
+withTxsInFailingBlock act = withTxsInFailingBlockM act . const . pure
+
+-- | Gather all the txs submitted by @act@ and resubmit them as a block that's expected to fail
+-- and compute the expected predicate failures from the created block using the supplied action.
+withTxsInFailingBlockM ::
+  ( HasCallStack
+  , ShelleyEraImp era
+  ) =>
+  ImpTestM era () ->
+  (Block BHeaderView era -> ImpTestM era (NonEmpty (PredicateFailure (EraRule "BBODY" era)))) ->
+  ImpTestM era ()
+withTxsInFailingBlockM act mkExpectedFailures = do
+  txs <- simulateThenRestore $ impTxsFrom act
+  submitFailingBlockM txs mkExpectedFailures
+
+-- | Submit a list of transactions as a block that's expected to succeed.
+-- The transactions are not modified to make them balance.
+submitBlock_ ::
+  ( HasCallStack
+  , ShelleyEraImp era
+  ) =>
+  [Tx TopTx era] -> ImpTestM era ()
+submitBlock_ = void . submitBlock
+
+-- | Submit a list of transactions as a block that's expected to succeed.
+-- The transactions are not modified to make them balance.
+submitBlock ::
+  ( HasCallStack
+  , ShelleyEraImp era
+  ) =>
+  [Tx TopTx era] -> ImpTestM era (Block BHeaderView era)
+submitBlock = expectRightDeepExpr . first fst <=< trySubmitBlock
+
+-- | Submit a list of transactions as a block that's expected to fail
+-- with the given predicate failures.
+-- The transactions are not modified to make them balance.
+submitFailingBlock ::
+  ( HasCallStack
+  , ShelleyEraImp era
+  ) =>
+  [Tx TopTx era] ->
+  NonEmpty (PredicateFailure (EraRule "BBODY" era)) ->
+  ImpTestM era ()
+submitFailingBlock txs = submitFailingBlockM txs . const . pure
+
+-- | Submit a list of transactions as a block that's expected to be rejected,
+-- and compute the expected predicate failures from the created block using the supplied action.
+-- The transactions are not modified to make them balance.
+submitFailingBlockM ::
+  ( HasCallStack
+  , ShelleyEraImp era
+  ) =>
+  [Tx TopTx era] ->
+  (Block BHeaderView era -> ImpTestM era (NonEmpty (PredicateFailure (EraRule "BBODY" era)))) ->
+  ImpTestM era ()
+submitFailingBlockM txs mkExpectedFailures = do
+  (predFailures, block) <- expectLeftDeepExpr =<< trySubmitBlock txs
+  expectedFailures <- mkExpectedFailures block
+  predFailures `shouldBeExpr` expectedFailures
+
+-- | Try to submit a list of transactions as a block,
+-- Return the block that was created using the transactions and any predicate
+-- failures that are produced.
+-- The transactions are not modified to make them balance.
+trySubmitBlock ::
```

Comment:
I think in order to have it consistent with `trySubmitTx`, this function should accept a list of transactions that have not been fixed up and this function itself will call the `simulateThenRestore` in order to do all the fixing up and then submission as a block.

If we don't do that then usability of these functions will be limited.